common/src/buffers/in_mem/recv.rs

```diff
@@ -0,0 +1,195 @@
+use async_trait::async_trait;
+use log::error;
+use log::info;
+use prost::Message as _;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt::Display;
+use std::hash::Hash;
+use std::mem::take;
+
+use crate::buffers::ChunkDecodeError;
+use crate::buffers::DenimChunk;
+use crate::buffers::Flag;
+use crate::buffers::ReceivingBuffer;
+use crate::denim_message::DeniableMessage;
+
+#[derive(Debug)]
+pub struct ChunkBuffer {
+    chunks: HashMap<u32, Vec<u8>>,
+    waiting_for: HashSet<u32>,
+}
+
+impl Default for ChunkBuffer {
+    fn default() -> Self {
+        let mut waiting_for = HashSet::new();
+        waiting_for.insert(0);
+        ChunkBuffer {
+            chunks: Default::default(),
+            waiting_for,
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct InMemoryReceivingBuffer<Sender: Eq + Hash> {
+    buffers: HashMap<Sender, HashMap<u32, ChunkBuffer>>,
+}
+
+#[async_trait]
+impl<T: Send + Eq + Hash + Copy + Display> ReceivingBuffer<T> for InMemoryReceivingBuffer<T> {
+    async fn process_chunks(
+        &mut self,
+        sender: T,
+        chunks: Vec<DenimChunk>,
+    ) -> Vec<Result<DeniableMessage, ChunkDecodeError>> {
+        let buffer = self.buffers.entry(sender).or_default();
+        let mut messages = Vec::new();
+        for mut chunk in chunks {
+            if chunk.flag() == Flag::DummyPadding {
+                continue;
+            }
+            let message_id = chunk.message_id();
+            let chunk_buffer = buffer.entry(message_id).or_default();
+            let seq = chunk.sequence_number();
+
+            if !chunk_buffer.waiting_for.contains(&seq) {
+                for id in 0..seq {
+                    if !chunk_buffer.chunks.contains_key(&id) {
+                        chunk_buffer.waiting_for.insert(id);
+                    }
+                }
+                let next = chunk.sequence_number() + 1;
+                if chunk.flag() != Flag::Final && !chunk_buffer.chunks.contains_key(&next) {
+                    chunk_buffer.waiting_for.insert(next);
+                }
+                info!(
+                    "Received message chunk out of order from sender {} for message {:?}. Waiting for {:?}",
+                    sender,
+                    chunk.message_id(),
+                    chunk_buffer.waiting_for
+                )
```

Comment:
```suggestion
                info!(
                    "Received message chunk {:?} out of order from sender {} for message {:?}. Waiting for {:?}",
                    sender,
                    chunk.sequence_number(),
                    chunk.message_id(),
                    chunk_buffer.waiting_for
                )
```

Reply:
Du skal bytte rundt på sender og sequence number, jeg har gjort det for dig 😘